java-checks/src/main/java/org/sonar/java/checks/RedundantTypeCastCheck.java

```diff
@@ -114,16 +115,10 @@ private static Type targetType(TypeCastTree typeCastTree) {
         return null;
       case VARIABLE:
         return ((VariableTree) parent).symbol().type();
-      case METHOD_INVOCATION:
-        // FIXME(SONARJAVA-3191) dead code
-        // If placed in invocation arguments, parent of typecast will be ARGUMENTS, not METHOD_INVOCATION
-        MethodInvocationTree mit = (MethodInvocationTree) parent;
-        if (mit.symbol().isMethodSymbol()) {
-          Symbol.MethodSymbol sym = (Symbol.MethodSymbol) mit.symbol();
-          int castArgIndex = mit.arguments().indexOf(typeCastTree);
-          return sym.parameterTypes().get(castArgIndex);
-        }
-        return null;
+      case ARGUMENTS:
+        Arguments arguments = (Arguments)parent;
+        int castArgIndex = arguments.indexOf(typeCastTree);
+        return arguments.get(castArgIndex).symbolType();
```

Comment:
Sorry I didn't catch it earlier... this won't work. What you wrote here is not equivalent to what we had earlier.
What is here now is this equivalent to do : `typeCastTree.symbolType()`, so the type of the resulting cast. We already have this information and that's not what we are searching for.

Previously, we were looking at the symbol of the method which was called, and looking at the EXPECTED type, as defined in his declaration.

In the following example: previously we would have retrieved `Collection`, today we are getting `ArrayList`

```
void foo(Collection param) { ... }
//       ^^^^^^^^^^ type from declaration = Collection

void bar(List param) {
  foo((ArrayList) param);
//     ^^^^^^^^^ type from cast = ArrayList
}
```

So in fact probably you will still need to get to the NewClassTree or MethodInvocationTree, but the symbol corresponding is only one parent ahead.

Reply:
You're totally right. Should be fixed